File: src/StockAdvisorFS/WorkflowSample.fs
module StockAdvisorFS.WorkflowSample

open System
open System.Threading.Tasks
open AgentNet
open Anthropic
open Microsoft.Extensions.AI

// ============================================================================
// WORKFLOW SAMPLE
// Demonstrates the AgentNet workflow DSL with parallel execution,
// sequential pipeline, and AI agent integration for comparing two stocks.
// ============================================================================

// ----------------------------------------------------------------------------
// Chat Client Setup
// ----------------------------------------------------------------------------

let private failIfNone msg opt = opt |> Option.defaultWith (fun () -> failwith msg)
let private tryGetEnv = Environment.GetEnvironmentVariable >> Option.ofObj

let private createChatClient () =
    let apiKey = tryGetEnv "ANTHROPIC_API_KEY" |> failIfNone "ANTHROPIC_API_KEY environment variable is not set."
    let model = tryGetEnv "ANTHROPIC_MODEL" |> Option.defaultValue "claude-sonnet-4-20250514"
    let client = AnthropicClient(APIKey = apiKey)
    client.AsIChatClient(model)

let private createAnalysisAgent (chatClient: IChatClient) =
    ChatAgent.create """
        You are a stock comparison analyst. Given information about two stocks,
        provide a concise comparison highlighting:
        - Which stock appears stronger and why
        - Key differences in their metrics
        - Risk considerations
        - A brief recommendation
        Keep your response focused and under 200 words.
        """
    |> ChatAgent.withName "StockAnalyst"
    |> ChatAgent.build chatClient

// ----------------------------------------------------------------------------
// Domain Types
// ----------------------------------------------------------------------------

type StockSymbols = { Symbol1: string; Symbol2: string }
type StockData = { Symbol: string; Info: string; Volatility: string }
type StockPair = { Stock1: StockData; Stock2: StockData }
type AnalysisResult = { Pair: StockPair; Analysis: string }

// ----------------------------------------------------------------------------
// Workflow Executors
// ----------------------------------------------------------------------------

/// Fetches stock info and volatility for a symbol
let private fetchStockData (symbol: string) : Task<StockData> = task {
    let! info = StockTools.getStockInfo symbol
    let! volatility = StockTools.calculateVolatility symbol
    return { Symbol = symbol; Info = info; Volatility = volatility }
}

/// Fetches data for both stocks in parallel
let private fetchBothStocks (symbols: StockSymbols) : Task<StockPair> = task {
    let! results = Task.WhenAll([|
        fetchStockData symbols.Symbol1
        fetchStockData symbols.Symbol2
    |])
    return { Stock1 = results.[0]; Stock2 = results.[1] }
}

/// Formats a StockPair into a prompt for the AI agent
let private formatStockPair (pair: StockPair) =
    $"""Compare these two stocks:

{pair.Stock1.Symbol}:
{pair.Stock1.Info}
Volatility: {pair.Stock1.Volatility}

{pair.Stock2.Symbol}:
{pair.Stock2.Info}
Volatility: {pair.Stock2.Volatility}"""

/// Parses the AI response into an AnalysisResult
let private parseAnalysisResult (pair: StockPair) (response: string) =
    { Pair = pair; Analysis = response }

/// Creates a typed agent for stock comparison
let private createTypedAnalysisAgent (agent: ChatAgent) =
    TypedAgent.create formatStockPair parseAnalysisResult agent    

/// Generates a formatted report
let private generateReport (result: AnalysisResult) : string =
    $"""
================================================================================
                        STOCK COMPARISON REPORT
================================================================================

--- {result.Pair.Stock1.Symbol} ---
{result.Pair.Stock1.Info}

{result.Pair.Stock1.Volatility}

--- {result.Pair.Stock2.Symbol} ---
{result.Pair.Stock2.Info}

{result.Pair.Stock2.Volatility}

--- AI ANALYSIS ---
{result.Analysis}
================================================================================
"""

// ----------------------------------------------------------------------------
// Public Entry Point
// ----------------------------------------------------------------------------

/// Runs the workflow sample comparing two stocks
let run (symbol1: string) (symbol2: string) : Task<unit> = task {
    printfn $"\nRunning comparison workflow for {symbol1} vs {symbol2}..."
    printfn "Initializing AI agent..."

    let chatClient = createChatClient ()
    let chatAgent = createAnalysisAgent chatClient
    let typedAgent = createTypedAnalysisAgent chatAgent

    let wf = workflow {
        step (Executor.fromTask "FetchBothStocks" fetchBothStocks)
        step (Executor.fromTypedAgent "CompareStocks" typedAgent)
        step (Executor.fromFn "GenerateReport" generateReport)
    }

    printfn "Step 1: Fetching stock data in parallel..."
    printfn "Step 2: AI agent analyzing stocks..."

    let input = { Symbol1 = symbol1; Symbol2 = symbol2 }
    let! report = Workflow.run input wf

    printfn "%s" report
}

File: src/StockAdvisorCS/WorkflowSample.cs
using Microsoft.Agents.AI.Workflows;

namespace StockAdvisorCS;

// ============================================================================
// WORKFLOW SAMPLE
// Demonstrates the MAF WorkflowBuilder with parallel execution and
// sequential pipeline for comparing two stocks.
// ============================================================================

// ----------------------------------------------------------------------------
// Domain Types
// ----------------------------------------------------------------------------

public record StockSymbols(string Symbol1, string Symbol2);
public record StockData(string Symbol, string Info, string Volatility);
public record StockPair(StockData Stock1, StockData Stock2);
public record ComparisonReport(StockPair Pair, string Comparison);

// ----------------------------------------------------------------------------
// Workflow Executors
// ----------------------------------------------------------------------------

internal sealed class FetchBothStocksExecutor() : Executor<StockSymbols, StockPair>("FetchBothStocks")
{
    public override async ValueTask<StockPair> HandleAsync(
        StockSymbols symbols,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        Console.WriteLine("  [FetchBothStocks] Fetching stock data in parallel...");

        var task1 = FetchStockDataAsync(symbols.Symbol1);
        var task2 = FetchStockDataAsync(symbols.Symbol2);
        await Task.WhenAll(task1, task2);

        return new StockPair(task1.Result, task2.Result);
    }

    private static async Task<StockData> FetchStockDataAsync(string symbol)
    {
        var info = await StockTools.GetStockInfo(symbol);
        var volatility = await StockTools.CalculateVolatility(symbol);
        return new StockData(symbol, info, volatility);
    }
}

internal sealed class CompareStocksExecutor() : Executor<StockPair, ComparisonReport>("CompareStocks")
{
    public override async ValueTask<ComparisonReport> HandleAsync(
        StockPair pair,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        Console.WriteLine("  [CompareStocks] Comparing stocks...");
        var comparison = await StockTools.CompareStocks(pair.Stock1.Symbol, pair.Stock2.Symbol);
        return new ComparisonReport(pair, comparison);
    }
}

internal sealed class GenerateReportExecutor() : Executor<ComparisonReport, string>("GenerateReport")
{
    public override ValueTask<string> HandleAsync(
        ComparisonReport report,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        Console.WriteLine("  [GenerateReport] Generating report...");
        var result = $"""

================================================================================
                        STOCK COMPARISON REPORT
================================================================================

--- {report.Pair.Stock1.Symbol} ---
{report.Pair.Stock1.Info}

{report.Pair.Stock1.Volatility}

--- {report.Pair.Stock2.Symbol} ---
{report.Pair.Stock2.Info}

{report.Pair.Stock2.Volatility}

--- COMPARISON ---
{report.Comparison}
================================================================================
""";
        return ValueTask.FromResult(result);
    }
}

// ----------------------------------------------------------------------------
// Workflow Sample
// ----------------------------------------------------------------------------

public static class WorkflowSample
{
    private static readonly Workflow _workflow = BuildWorkflow();

    private static Workflow BuildWorkflow()
    {
        // Create executors
        var fetchBothStocks = new FetchBothStocksExecutor();
        var compareStockPair = new CompareStocksExecutor();
        var reportExecutor = new GenerateReportExecutor();

        // Build the workflow using MAF WorkflowBuilder
        WorkflowBuilder builder = new(fetchBothStocks);
        builder.AddEdge(fetchBothStocks, compareStockPair);
        builder.AddEdge(compareStockPair, reportExecutor);
        builder.WithOutputFrom(reportExecutor);
        return builder.Build();
    }

    /// <summary>
    /// Runs the workflow sample comparing two stocks.
    /// </summary>
    public static async Task RunAsync(string symbol1, string symbol2)
    {
        Console.WriteLine($"\nRunning MAF workflow for {symbol1} vs {symbol2}...");

        var input = new StockSymbols(symbol1, symbol2);

        await using var run = await InProcessExecution.RunAsync(_workflow, input);

        // Process workflow events
        foreach (var evt in run.NewEvents)
        {
            if (evt is ExecutorCompletedEvent completed && completed.ExecutorId == "GenerateReport")
            {
                Console.WriteLine(completed.Data);
            }
        }
    }
}
